### PR TITLE
Validate that issuerCertificate is a sensible CA certificate

### PR DESCRIPTION
**Pull request by bartonjs**
When a certificate is given as a CA cert it needs to
* Not have a KeyUsage extension, or assert KeyCertSign.
* Have BasicConstraints.CA set or be self-signed and not have BasicConstraints.

These two rules are a discovered state of Windows chain validation after
cross-reading RFC 3280, RFC 5280 and ITU-T X.509 2012/10.

The "self-signed" portion turns out to be difficult, since both Windows and
OpenSSL take the AuthorityKeyIdentifier extension into account. It's even more difficult in that they don't seem to take them into account in the same
way.

So the existing two IsSelfSigned methods are deleted, and replaced with a
new one that takes the KeyId form of AKID into account, because they treat
that form the same.

We can be confused by the signature not working out, but that just means
we accept certs for signing that the chain builders reject. We will also believe
a certificate which uses the Name+Serial form of AuthorityKeyIdentifier is
not self-signed, so we will force it to have a Basic Constraints extension.

In addition to the CA cert sanity checks, enforce that the notBefore and
notAfter values supplied are aligned with the issuerCertificate's NotBefore
and NotAfter values.

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs [hunk 1]
```diff
@@ -158,6 +159,80 @@ internal static void DisposeAll(this IEnumerable<IDisposable> disposables)
...
+                }
+                else if (extension.Oid.Value == Oids.AuthorityKeyIdentifier)
+                {
+                    akid = new X509AuthorityKeyIdentifierExtension();
+                    akid.CopyFrom(extension);
+                }
+
+                if (akid != null && skid != null)
+                {
+                    break;
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: We should probably handle multiple copies of extensions (perhaps by throwing)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs [hunk 2]
```diff
@@ -158,6 +159,80 @@ internal static void DisposeAll(this IEnumerable<IDisposable> disposables)
...
+        }
+
+        internal static bool IsSelfSigned(
+            this X509Certificate2 cert,
+            X509SubjectKeyIdentifierExtension skid,
+            X509AuthorityKeyIdentifierExtension akid)
+        {
+            Debug.Assert(cert != null);
+
+            if (!cert.SubjectName.RawData.ContentsEqual(cert.IssuerName.RawData))
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: Do the subjects actually have to be byte-for-byte equal? For example, could I switch the parts of a DN and have an equivalent name?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs [hunk 3]
```diff
@@ -158,6 +159,80 @@ internal static void DisposeAll(this IEnumerable<IDisposable> disposables)
...
+            if (akid == null)
+            {
+                return true;
+            }
+
+            if (akid.KeyIdentifier != null)
+            {
+                if (skid == null)
+                {
+                    return false;
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: I believe we found a case like this on one of the Microsoft roots

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs [hunk 4]
```diff
@@ -158,6 +159,80 @@ internal static void DisposeAll(this IEnumerable<IDisposable> disposables)
...
+                }
+
+                if (!akid.KeyIdentifier.Equals(skid.SubjectKeyIdentifier, StringComparison.Ordinal))
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: You can shorten this to akid.KeyIdentifier != skid.SubjectKeyIdentifier

File: src/System.Security.Cryptography.X509Certificates/src/Resources/Strings.resx [hunk 5]
```diff
@@ -115,9 +115,21 @@
...
+    <value>The issuer certificate does not have an appropriate value for the Basic Constraints extension.</value>
+  </data>
+  <data name="Cryptography_CertReq_IssuerKeyUsageInvalid" xml:space="preserve">
+    <value>The issuer certificate's Key Usage extension is present but does not contain the KeyCertSign flag.</value>
+  </data>
   <data name="Cryptography_CertReq_IssuerRequiresPrivateKey" xml:space="preserve">
     <value>The provided issuer certificate does not have an associated private key.</value>
   </data>
+  <data name="Cryptography_CertReq_NotAfterNotNested" xml:space="preserve">
+    <value>The requested notAfter value '{0}' is later than issuerCertificate.NotAfter ({1}).</value>
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/Resources/Strings.resx]: You have the first date in quotes and the second in parens in both of these strings

File: src/System.Security.Cryptography.X509Certificates/src/Resources/Strings.resx [hunk 6]
```diff
@@ -115,9 +115,21 @@
...
+  </data>
+  <data name="Cryptography_CertReq_NotBeforeNotNested" xml:space="preserve">
+    <value>The requested notBefore value '{0}' is earlier than issuerCertificate.NotAFter ({1}).</value>
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/Resources/Strings.resx]: Capitalization in NotAfter (no need to capitalize the F)

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs [hunk 7]
```diff
@@ -365,6 +365,85 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
...
+                {
+                    akid = new X509AuthorityKeyIdentifierExtension();
+                    akid.CopyFrom(extension);
+                }
+
+                // Note that skid and akid are not present here.
+                // If we find the basic constraints extension we don't need the other two.
+                if (keyUsage != null && basicConstraints != null)
+                {
+                    break;
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs]: Same comment on multiple copies of extensions

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509AuthorityKeyIdentifierExtension.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.X509Certificates
+{
+    internal sealed class X509AuthorityKeyIdentifierExtension : X509Extension
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509AuthorityKeyIdentifierExtension.cs]: Should we consider making this public?

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509AuthorityKeyIdentifierExtension.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            get
+            {
+                if (!_decoded)
+                    Decode();
+
+                return _keyIdentifier;
+            }
+        }
+
+        internal bool HasComplexValue
```
Comment (morganbr) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509AuthorityKeyIdentifierExtension.cs]: This could use a comment since complex currently means @bartonjs thinks it's complicated

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509AuthorityKeyIdentifierExtension.cs (hunk 8 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509AuthorityKeyIdentifierExtension.cs]: Not with the abbreviated support it has, no.  Next release, hopefully.

**Comment by bartonjs** [comment]
Apparently Apple says Basic Constraints are always required.  Hmm.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs [hunk 10]
```diff
@@ -365,6 +365,85 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
...
+                    SR.Format(
+                        SR.Cryptography_CertReq_NotBeforeNotNested,
+                        notBeforeLocal,
+                        issuerCertificate.NotBefore),
+                    nameof(notBefore));
+            }
+
+            DateTime notAfterLocal = notAfter.LocalDateTime;
+
+            // Round down to the second, since that's the cert accuracy.
```
Comment (steveharter) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs]: So notbefore is not rounded up\down to the second? Is this actually overkill w.r.t. sub-second accuracy?

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs [hunk 11]
```diff
@@ -365,6 +365,85 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
...
+                {
+                    break;
+                }
+            }
+
+            if (basicConstraints != null && !basicConstraints.CertificateAuthority)
+                throw new ArgumentException(SR.Cryptography_CertReq_IssuerBasicConstraintsInvalid, nameof(issuerCertificate));
+
+            // Windows only allows Basic Constraints to be missing for a self-signed cert.
+            if (basicConstraints == null && !issuerCertificate.IsSelfSigned(skid, akid))
```
Comment (steveharter) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs]: Assert or check for null `skid`, `akid` here? Or just let NullReferenceException?

Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs]: `null` is a fine value to pass as a parameter... the IsSelfSigned method handles them being null.  Am I missing something in the comment?

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs (hunk 10 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs]: Certificates have one-second accuracy (it's stored as yyMMddHHmmss'Z', or yyyyMMddHHmmss'Z').

As the comment in the code suggests, the problem is the following code:

```C#
DateTimeOffset now = DateTimeOffset.UtcNow;
DateTimeOffset notAfter = now.AddDays(90);

using (var root = MakeRootCert(now, notAfter))
{
    ...
    using (var issued = certReq.Create(root, now, notAfter, serial))
    {
        ...
    }
}
```

The notAfter almost certainly contains some sub-second value, and the first one got rounded down (by the nature of certs), so now it's comparing 2017-04-24T15:35:21.333 to 2017-04-24T15:35:21.  Once it goes into a cert they're the same thing, so it shouldn't fail.

**Comment by bartonjs** [comment]
So, @morganbr; any thoughts about always requiring Basic Constraints + isCA=true?  It does match the RFC 3280 comments about the KeyCertSign bit (https://tools.ietf.org/html/rfc3280#section-4.2.1.3):

> The keyCertSign bit is asserted when the subject public key is
used for verifying a signature on public key certificates.  If the
keyCertSign bit is asserted, then the cA bit in the basic
constraints extension (section 4.2.1.10) MUST also be asserted.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: https://tools.ietf.org/html/rfc3280#section-9
> The X.500 series of specifications defines rules for
comparing distinguished names that require comparison of strings
without regard to case, character set, multi-character white space
substring, or leading and trailing white space.  This specification
relaxes these requirements, requiring support for binary comparison
at a minimum.

> CAs MUST encode the distinguished name in the subject field of a CA
certificate identically to the distinguished name in the issuer field
in certificates issued by that CA.  If CAs use different encodings,
implementations might fail to recognize name chains for paths that
include this certificate.  As a consequence, valid paths could be
rejected.

https://tools.ietf.org/html/rfc5280#section-7.1
> Two distinguished names DN1 and DN2 match if they
have the same number of RDNs, for each RDN in DN1 there is a matching
RDN in DN2, and the matching RDNs appear in the same order in both
DNs.

ITU T-REC X.509 says it follows the `distinguishedNameMatch` rule, which does a long-winded way of saying `binary match`, since it's dealing with it at concepts instead of encoded values.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs (hunk 3 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: That particular certificate also specified the Issuer+SerialNumber version, so it would have failed here anyways.  But if we decide to adopt BasicConstraints:isCA=true as a minbar we don't need to have IsSelfSigned, and we can revert this until we can figure out the best way to solve it.

File: src/System.Security.Cryptography.X509Certificates/src/Resources/Strings.resx (hunk 6 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Resources/Strings.resx]: Well, that and it should have said NotBefore :)

**Comment by bartonjs** [comment]
I'm seeing some unexpected results locally with the IsSelfSigned change.  Since that's not involved in the tree anymore I'll split it out and see if I'm comfortable making that improvement for 2.0 since it's mostly already ready (and should better align us in a few edge cases with the underlying systems).
